packages/modules/fulfillment/src/models/service-zone.ts

```diff
@@ -1,126 +1,60 @@
-  metadata: Record<string, unknown> | null = null
-
-  @ManyToOne(() => FulfillmentSet, {
-    type: "text",
-    mapToPk: true,
-    fieldName: "fulfillment_set_id",
-    onDelete: "cascade",
-  })
-  @FulfillmentSetIdIndex.MikroORMIndex()
-  fulfillment_set_id: string
-
-  @ManyToOne(() => FulfillmentSet, { persist: false })
-  fulfillment_set: Rel<FulfillmentSet>
+import { FulfillmentSet } from "./fulfillment-set"
+import { GeoZone } from "./geo-zone"
+import { ShippingOption } from "./shipping-option"
+
+export type ServiceZoneSchema = {
+  id: PrimaryKeyModifier<string, IdProperty>
+  name: TextProperty
+  fulfillment_set: BelongsTo<() => typeof FulfillmentSet>
+  geo_zones: HasMany<() => typeof GeoZone>
+  shipping_options: HasMany<() => typeof ShippingOption>
+  metadata: NullableModifier<Record<string, unknown>, JSONProperty>
+}
 
-  @OneToMany(() => GeoZone, "service_zone", {
-    cascade: [Cascade.PERSIST, "soft-remove"] as any,
-    orphanRemoval: true,
+export const ServiceZone = model
+  .define("service_zone", {
+    id: model.id({ prefix: "serzo" }).primaryKey(),
+    name: model.text(),
+    fulfillment_set: model.belongsTo<() => typeof FulfillmentSet>(
+      () => FulfillmentSet,
+      {
+        mappedBy: "service_zones",
+      }
+    ),
+    geo_zones: model.hasMany<() => typeof GeoZone>(() => GeoZone, {
+      mappedBy: "service_zone",
+    }),
+    shipping_options: model.hasMany<() => typeof ShippingOption>(
+      () => ShippingOption,
+      {
+        mappedBy: "service_zone",
+      }
+    ),
+    metadata: model.json().nullable(),
   })
-  geo_zones = new Collection<Rel<GeoZone>>(this)
-
-  @OneToMany(
-    () => ShippingOption,
-    (shippingOption) => shippingOption.service_zone,
+  .indexes([
     {
-      cascade: [Cascade.PERSIST, "soft-remove"] as any,
-      orphanRemoval: true,
-    }
-  )
-  shipping_options = new Collection<Rel<ShippingOption>>(this)
-
-  @Property({
-    onCreate: () => new Date(),
-    columnType: "timestamptz",
-    defaultRaw: "now()",
-  })
-  created_at: Date
-
-  @Property({
-    onCreate: () => new Date(),
-    onUpdate: () => new Date(),
-    columnType: "timestamptz",
-    defaultRaw: "now()",
-  })
-  updated_at: Date
-
-  @Index({
-    name: deletedAtIndexName,
-    expression: deletedAtIndexStatement,
-  })
-  @Property({ columnType: "timestamptz", nullable: true })
-  deleted_at: Date | null = null
-
-  @BeforeCreate()
-  onCreate() {
-    this.id = generateEntityId(this.id, "serzo")
-    this.fulfillment_set_id ??= this.fulfillment_set?.id
-  }
-
-  @OnInit()
-  onInit() {
-    this.id = generateEntityId(this.id, "serzo")
-    this.fulfillment_set_id ??= this.fulfillment_set?.id
-  }
-}
+      on: ["name"],
+      unique: true,
+      where: "deleted_at IS NULL",
+    },
+  ])
+  .cascades({
+    delete: ["geo_zones", "shipping_options"],
+  }) as unknown as DmlEntity<
+  DMLEntitySchemaBuilder<ServiceZoneSchema>,
+  "ServiceZone"
+>
```

Comment:
When there is too many inferences typescript reach the maximum issue and you have to provide a type yourself in the middle to reduce the complexity of the overall inference. Otherwise you have to ignore and completely loose typings which defeats the purpose

Reply:
Cool, thx for explaining.